src/main/java/com/sequenceiq/cloudbreak/service/stack/connector/aws/AwsConnector.java

```diff
@@ -177,6 +163,62 @@ public void buildStack(Stack stack, String userData, Map<String, Object> setupPr
         }
     }
 
+    private String getEbsSnapshotIdIfNeeded(Stack stack, Long stackId) {
+        if (isEncryptedVolumeRequested(stack)) {
+            Optional<String> snapshot = createSnapshotIfNeeded(stack);
+            if (snapshot.isPresent()) {
+                return snapshot.orNull();
+            } else {
+                throw new CloudFormationStackException(String.format("Failed to create Ebs encrypted volume on stack: %s", stackId));
+            }
+        } else {
+            return null;
+        }
+    }
+
+    private Optional<String> createSnapshotIfNeeded(Stack stack) {
+        AmazonEC2Client client = awsStackUtil.createEC2Client(stack);
+        DescribeSnapshotsRequest describeSnapshotsRequest = new DescribeSnapshotsRequest()
+                .withFilters(
+                        new Filter().withName("tag-key").withValues(CLOUDBREAK_EBS_SNAPSHOT),
+                        new Filter().withName("tag-value").withValues(CLOUDBREAK_EBS_SNAPSHOT));
+        DescribeSnapshotsResult describeSnapshotsResult = client.describeSnapshots(describeSnapshotsRequest);
+        if (describeSnapshotsResult.getSnapshots().isEmpty()) {
+            DescribeAvailabilityZonesResult availabilityZonesResult = client.describeAvailabilityZones(new DescribeAvailabilityZonesRequest()
+                    .withFilters(new Filter().withName("region-name").withValues(Regions.valueOf(stack.getRegion()).getName())));
+            CreateVolumeResult volumeResult = client.createVolume(new CreateVolumeRequest()
+                    .withSize(SNAPSHOT_VOLUME_SIZE)
+                    .withAvailabilityZone(availabilityZonesResult.getAvailabilityZones().get(0).getZoneName())
+                    .withEncrypted(true));
+            EbsVolumeStatePollerObject ebsVolumeStateObject =
+                    new EbsVolumeStatePollerObject(stack, volumeResult, volumeResult.getVolume().getVolumeId(), client);
+            PollingResult pollingResult = ebsVolumeStatePollerObjectPollingService
+                    .pollWithTimeout(ebsVolumeStateCheckerTask, ebsVolumeStateObject, POLLING_INTERVAL, INFINITE_ATTEMPTS);
+            if (PollingResult.isSuccess(pollingResult)) {
+                CreateSnapshotResult snapshotResult = client.createSnapshot(new CreateSnapshotRequest()
+                        .withVolumeId(volumeResult.getVolume().getVolumeId())
+                        .withDescription("Encrypted snapshot"));
+                CreateTagsRequest createTagsRequest = new CreateTagsRequest()
+                        .withTags(ImmutableList.of(new Tag().withKey(CLOUDBREAK_EBS_SNAPSHOT).withValue(CLOUDBREAK_EBS_SNAPSHOT)))
+                        .withResources(snapshotResult.getSnapshot().getSnapshotId());
+                client.createTags(createTagsRequest);
```

Comment:
I think the snapshot creation is usually done in a second as it is an empty 10GB disk but if it's stuck for some reason we'll have to wait until it is ready, so a poller is needed here as well to check the snapshot progress/state.